### PR TITLE
solution-level props and targets

### PR DESCRIPTION
**Pull request by baronfel**
Sample of using a solution-level targets file to replace some of the fantomas scripts. Notes/docs may be more forthcoming. Use them like `dotnet build -t:<Target name>`

### CODE REVIEW

File: Directory.Solution.targets [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    <Target Name="EnsureRepoConfig">
+        <Exec Command="git config core.hooksPath .githooks" />
+    </Target>
+
+    <Target Name="Benchmark">
+        <Exec Command="dotnet build -c Release src/Fantomas.Benchmarks" />
+        <Exec Command="dotnet src/Fantomas.Benchmarks/bin/Release/net6.0/Fantomas.Benchmarks.dll" IgnoreStandardErrorWarningFormat="true" />
+    </Target>
+    
+    <Target Name="All">
```
Comment (nojaf) [inline Directory.Solution.targets:71]: How acceptable does this look like?
It is what we do today with FAKE but so much more straightforward 😱

**Review by nojaf — CHANGES_REQUESTED** [review]
This is starting to grow on me.
I'm doing the request changes thing so it doesn't accidentally get merged.
I have a couple of questions left for you @baronfel.

File: Directory.Solution.targets [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                Command="dotnet fantomas @(FormatItems, ' ')"
+                Condition="@(FormatItems->Count()) > 0" />
+    </Target>
+
+    <Target Name="EnsureRepoConfig">
+        <Exec Command="git config core.hooksPath .githooks" />
+    </Target>
+
+    <Target Name="Benchmark">
+        <Exec Command="dotnet src/Fantomas.Benchmarks/bin/Release/net6.0/Fantomas.Benchmarks.dll" IgnoreStandardErrorWarningFormat="true" />
```
[the remark this answers is not in the document]
Comment (baronfel) [reply Directory.Solution.targets:68]: There's a way to invoke a specific target on a specific project with specific parameters that I was going to look at today. If I did that, then I'd say 'invoke the Test target on the Benchmarks project using Framework X and Configuration Y', and that would implicitly perform the build if necessary

**Review by baronfel — COMMENTED** [review]
General notes

File: Directory.Solution.targets (hunk 2 above)
[the remark this answers is not in the document]
Comment (nojaf) [reply Directory.Solution.targets:68]: Yes, that same to mind as well. Anything `dotnet command` will eventually call some target somewhere right? So all those commands can probably be replaced with target invokations.

[the remark this answers is not in the document]
Comment (baronfel) [reply Directory.Solution.targets:68]: Yep! And by the same token, if you needed the path to an output dll for some reason it's better to get the project reference for that project and ask it for a particular property, then use that property. These are all things that are _doable_, it's a matter of making them ergonomic.

[the remark this answers is not in the document]
Comment (baronfel) [reply Directory.Solution.targets]: I was able to do this pretty easily - the `OutputItems` of the `Build` target are the built assemblies, so I could build the benchmarks project and get that dll to execute (the build doesn't actually rerun if it has already happened once).

File: Directory.Solution.targets (hunk 1 above)
Comment (baronfel) [reply Directory.Solution.targets:75]: I got this working pretty well locally - calling targets on the solution is done through the `MSBuild` task.

**Comment by nojaf** [comment]
@baronfel I'm getting the same `C:\Users\nojaf\Projects\fantomas\fantomas.sln.metaproj : error MSB4057: The target "VSTest" does not exist in the project. [C:\Users\nojaf\Projects\fantomas\fantomas.sln]` error locally.

Maybe `dotnet test fantomas.sln` does a bit of orchestration and calls `VSTest` for all the projects it finds.

Changing this to
```xml
        <MSBuild
                Projects="src/Fantomas.Core.Tests/Fantomas.Core.Tests.fsproj"
                Targets="VSTest"
                Properties="Configuration=Release" />
        <MSBuild Projects="src/Fantomas.Tests/Fantomas.Tests.fsproj"   
                 Targets="VSTest"
                 Properties="Configuration=Release" />
```

Only runs the tests of the `Core` projects, very weird.

**Comment by nojaf** [comment]
Closing in favour of https://github.com/fsprojects/fantomas/pull/2531.

**Pull request closed without merging**